docs/source/developer/inlining.rst

```diff
@@ -76,10 +76,41 @@ The ``inline`` keyword argument can be one of three values:
   In all cases the function should return True to inline and return False to not
   inline, this essentially permitting custom inlining rules (typical use might
   be cost models).
+* Recursive functions with ``inline='always'`` will result in a non-terminating
+  compilation. If you wish to avoid this, use a cost model.
 
 .. note:: No guarantee is made about the order in which functions are assessed
           for inlining or about the order in which they are inlined.
 
+Inlining cost models
+====================
+When using recursive inlines, you can terminate the compilation by using
+a cost model.
+
+.. code:: python
+
+    from numba import njit
+    import numpy as np
+
+    class CostModel(object):
+
+        def __init__(self, max_inlines):
+            self._count = 0
+            self._max_inlines = max_inlines
+
+        def __call__(self, expr, caller, callee):
+            ret = self._count < self._max_inlines
+            self._count +=1
+            return ret
+
+    @njit(inline=CostModel(3))
+    def factorial(n):
+        if n <= 0:
+            return 1
+        return n * factorial(n - 1)
+
+    factorial(5)
+
```

Comment:
I feel that this section should go at the end of the document and be called something like "Using a function to limit the inlining depth of a recursive function". The two sections beneath here introduce the concepts more generally.